### ISSUE TITLE
0.14: Broken state file compatibility: RFC3339 Timestamp Format errors in aws_s3_bucket

### ISSUE THREAD

**Issue by phyber** [hashicorp/terraform#27103]
### Terraform Version

```
Terraform v0.14.0
+ provider registry.terraform.io/hashicorp/aws v3.19.0
```

### Terraform Configuration Files

```terraform
resource aws_s3_bucket bucket {
  for_each = local.machines

  bucket = each.key
  acl    = "private"
  tags   = each.value.tags

  server_side_encryption_configuration {
    rule {
      apply_server_side_encryption_by_default {
        sse_algorithm = "AES256"
      }
    }
  }

  # We have to ignore this as Arq manages its own lifecycle policies.
  lifecycle {
    ignore_changes = [
      lifecycle_rule,
    ]
  }
}
```

### Debug Output

Edit: Added this small bit of debug output to show AWS returning correct data.

```
2020-12-03T10:59:29.417Z [DEBUG] plugin.terraform-provider-aws_v3.19.0_x5: 2020/12/03 10:59:29 [DEBUG] S3 bucket: the-bucket-name, read lifecycle rule: {
2020-12-03T10:59:29.417Z [DEBUG] plugin.terraform-provider-aws_v3.19.0_x5:   ID: "Arq S3Glacier Lifecycle Policy DO NOT MODIFY",
2020-12-03T10:59:29.417Z [DEBUG] plugin.terraform-provider-aws_v3.19.0_x5:   Prefix: "glacier/",
2020-12-03T10:59:29.417Z [DEBUG] plugin.terraform-provider-aws_v3.19.0_x5:   Status: "Enabled",
2020-12-03T10:59:29.417Z [DEBUG] plugin.terraform-provider-aws_v3.19.0_x5:   Transitions: [{
2020-12-03T10:59:29.417Z [DEBUG] plugin.terraform-provider-aws_v3.19.0_x5:       Days: 0,
2020-12-03T10:59:29.417Z [DEBUG] plugin.terraform-provider-aws_v3.19.0_x5:       StorageClass: "GLACIER"
2020-12-03T10:59:29.417Z [DEBUG] plugin.terraform-provider-aws_v3.19.0_x5:     }]
2020-12-03T10:59:29.417Z [DEBUG] plugin.terraform-provider-aws_v3.19.0_x5: }
```

A small snippet showing that AWS seems to be returning the correct information, there is no `""` set in the `date` field here. (See: Additional Context)

### Expected Behavior

I should have been able to execute a `terraform plan` on 0.14.0 with my 0.13.5 state and have it work with no changes.

### Actual Behavior

The following error is produced:

```
Error: "" cannot be parsed as RFC3339 Timestamp Format

  on modules/backups/bucket.tf line 2, in resource "aws_s3_bucket" "bucket":
   2: resource aws_s3_bucket bucket {
```

### Steps to Reproduce

`terraform plan` after updating to 0.14.0.

### Additional Context

If I had to guess, this is happening due to the ignored lifecycle rules. Looking at those rules in the tfstate shows the following:

```json
            "lifecycle_rule": [
              {
                "abort_incomplete_multipart_upload_days": 0,
                "enabled": true,
                "expiration": [],
                "id": "Arq S3Glacier Lifecycle Policy DO NOT MODIFY",
                "noncurrent_version_expiration": [],
                "noncurrent_version_transition": [],
                "prefix": "glacier/",
                "tags": {},
                "transition": [
                  {
                    "date": "",
                    "days": 0,
                    "storage_class": "GLACIER"
                  }
                ]
              },
              {
                "abort_incomplete_multipart_upload_days": 0,
                "enabled": true,
                "expiration": [],
                "id": "Arq IA Lifecycle DO NOT MODIFY <a UUID was here>",
                "noncurrent_version_expiration": [],
                "noncurrent_version_transition": [],
                "prefix": "<a UUID was here>/",
                "tags": {},
                "transition": [
                  {
                    "date": "",
                    "days": 30,
                    "storage_class": "STANDARD_IA"
                  }
                ]
              }
            ],
```

The transition `date` is set to `""` in all of the buckets that show this error.

**Comment by jbardin** [comment]
Hi @phyber 

Thanks for filing the issue. This particular case was supposed to be caught by config validation, but the change was too broad to get into 0.14. The `lifecycle_rule` block is not declared in your configuration, and because `ignore_changes` is only meant to prevent changes from the configuration it cannot operate on the value correctly.

Since the configuration is not intended to change `lifecycle_rule` anyway, can you help me understand what behavior you are trying to prevent by setting that in `ignore_changes`? 

Thanks!

**Comment by phyber** [comment]
Hi, thanks for the quick response.

The ignore changes here are because the third party (closed source) software that is interacting with the bucket manages its own lifecycle policies. This behaviour cannot be changed.

The use of the bucket is something along the lines of:
  - Start with Terraform 0.12
  - Write some Terraform to setup buckets and surrounding resources
  - Give the software some access keys to interact with the bucket
  - The software starts reading/writing from/to the bucket, and also sets up some lifetime policies
    - These policies are unknown to Terraform, and may change depending on the software configuration
  - Re-run Terraform, the resource's ignored lifecycle rules end up in the state
    - This could have happened at any point from 0.12 until now
    - Multiple Terraform versions and AWS provider versions have been invoked during this time
    - Appears to be an intended behaviour, `TF_LOG=trace` shows Terraform describing Lifecycle state even though it's ignored
  - Re-run Terraform on 0.14, Terraform errors on state that is ignored

At this point, if I were to re-run Terraform without the ignore, Terraform would remove the unknown lifecycle policies and upset the third party software. This is actually difficult to recover from and isn't as simple as manually recreating the policies.

> The `lifecycle_rule` block is not declared in your configuration, and because `ignore_changes` is only meant to prevent changes from the configuration it cannot operate on the value correctly.

If I'm reading this correctly, you're saying that I'm using `ignore_changes` incorrectly, because `lifecycle_rule` was never declared in Terraform. If that is so, I'll have to disagree based on the Terraform documentation.

> In some rare cases, settings of a remote object are modified by processes outside of Terraform, which Terraform would then attempt to "fix" on the next run. In order to make Terraform share management responsibilities of a single object with a separate process, the `ignore_changes` meta-argument specifies resource attributes that Terraform should ignore when planning updates to the associated remote object.

The use of `ignored_changes` here appears to be an accepted use when an outside process may change a resource configuration outside of Terraform.

At this point, I believe I have enough information to manually fix the state, either by removing the broken `date` field from each of the transitions its listed in, or just removing the `lifecycle_rule` state entirely and letting Terraform re-add it. However, I'm keeping it around for now in case you're able to modify this behaviour in Terraform.

**Comment by jbardin** [comment]
Hi @phyber, thanks for the added info. The documentation you quoted is correct, and `ignore_changes` is still going to be a supported method to avoid "drift" from external changes. What I'm referring to is that if `lifecycle_rule` is computed and not set in the configuration, terraform should not attempt to change it in the first place regardless of `ignore_changes`. 

It appears the problem is that the provider is not setting the values correctly, or part of the schema is not defined correctly. 
We may be able to just avoid the issue with whatever fix is applied for #27106, if this is only a problem with the provider validating the fields. If that's the case we could continue using `ignore_change` to work around the bug, but it would be good to pinpoint what is causing the change to happen at all.

**Comment by phyber** [comment]
The schema for the transition in this case is:

```go
"transition": {
        Type:     schema.TypeSet,
        Optional: true,
        Set:      transitionHash,
        Elem: &schema.Resource{
                Schema: map[string]*schema.Schema{
                        "date": {
                                Type:         schema.TypeString,
                                Optional:     true,
                                ValidateFunc: validateS3BucketLifecycleTimestamp,
                        },
                        "days": {
                                Type:         schema.TypeInt,
                                Optional:     true,
                                ValidateFunc: validation.IntAtLeast(0),
                        },
                        "storage_class": {
                                Type:         schema.TypeString,
                                Required:     true,
                                ValidateFunc: validateS3BucketLifecycleTransitionStorageClass(),
                        },
                },
        },
},
```

With a validator of:

```go
func validateS3BucketLifecycleTimestamp(v interface{}, k string) (ws []string, errors []error) {
        value := v.(string)
        _, err := time.Parse(time.RFC3339, fmt.Sprintf("%sT00:00:00Z", value))
        if err != nil {
                errors = append(errors, fmt.Errorf(
                        "%q cannot be parsed as RFC3339 Timestamp Format", value))
        }

        return
}
```

Perhaps in some past version of Terraform an empty string was treated differently here. The relevant code in the provider doesn't seem to have changed in quite some time.

My guess for how this happened is:
  - Some older version of Terraform wrote these empty optional fields to the state
    - Perhaps newer versions of Terraform avoid doing this, it should be simple to test at some point
    - Maybe along the way, some Terraform version was supposed to clean up these values, but didn't because of `ignore_changes`
  - When reading the state back, Terraform, knowing they were optional and set to the type's (`string`) default value (an empty string in the case of Go) somehow hid them from the provider, so there was nothing to validate
  - Terraform 0.14 has changed behaviour in this area, so now the empty string is passed to the provider during planning, triggering the validation and erroring out

I'd check on this, but I'm unsure where this happens in the code.
